### ISSUE TITLE
Getting started: attrs not supported

### ISSUE THREAD

**Issue by ayroblu** [microsoft/pyright#146]
Hi, I wanted to try this out after mypy, but I don't understand some of the errors I'm getting.

## First
This looks like an attr.s thing
```
  10:17 - error: Expected no arguments to 'Foo' constructor
  10:17 - error: 'dq.Foo(dq.Bar.Thing.value)' has type 'Type[Foo]' and is not callable
```

Declaration:
```python
import attr
from enum import Enum

@attr.s(frozen=True)
class Foo:
    name: str = attr.ib()

class Bar(Enum):
    Thing = "Thing"

```

Usage:
```python
    a=dq.Foo(dq.Bar.Thing.value),
```

## Second
This looks like it doesn't like @classmethod? Or it's an attr.s thing again
```
  45:6 - error: Argument of type '(cls: Type[Bar], details: Dict[str, Any]) -> Rule' cannot be assigned to parameter 'f' of type 'function'
```

```python
@attr.s(auto_attribs=True, frozen=True)
class Foo(Rule):
    @classmethod # this line
    def _deserialize(cls, details: Dict[str, Any]) -> Rule:
        pass
```

**Comment by erictraut** [comment]
The "attr" class decorator is modifying the standard behavior of the class it's decorating, and it's doing so in a way that the type checker has no visibility. If you look at the undecorated class Foo, there's no `__init__` method specified and it derives from no other object that specifies an `__init__` method, so pyright assumes that its initializer takes no parameters. The type stub file that accompanies the attires module contains a type definition for the `@attr.s` class decorator. This type definition indicates that it returns `Callable[[_C], _C]`, which is to say that the `__init__` method is unmodified by the decorator.

Class decorators that do "clever" things with python are difficult for type checkers to deal with in general because the type annotation language isn't rich enough to express the arbitrary behavioral changes that they can make to the class and its members.

I'm not sure what to suggest here other than stop using attrs if you want to use type checking.

**Comment by ayroblu** [comment]
Thanks so much @erictraut I'll stick with MyPy

**Comment by Sieboldianus** [comment]
It would be nice if we could use

 ```python
import attr # type: ignore
```
to disable all pyright checks for this package.

**Comment by alukach** [comment]
I don't believe this issue should be closed.

For the record, [`dataclasses.dataclass`](https://docs.python.org/3.7/library/dataclasses.html#dataclasses.dataclass) does work with `pyright` and (from the end-user's perspective) does the same thing (not sure about the differences under-the-hood).

**Comment by erictraut** [comment]
The difference is that dataclass is part of the official python spec, whereas attr is not. Pyright contains a bunch of code specifically to handle dataclass. It does not contain code to deal with special-case behaviors implemented by arbitrary third-party libraries. I recommend moving away from attr to dataclass if it meets your needs.

**Comment by Sieboldianus** [comment]
> The difference is that dataclass is part of the official python spec, whereas attr is not. Pyright contains a bunch of code specifically to handle dataclass. It does not contain code to deal with special-case behaviors implemented by arbitrary third-party libraries. I recommend moving away from attr to dataclass if it meets your needs.

That is exactly what I ended up doing, and I don't regret it (except that my minimum dependeny is now pyton 3.7)

**Comment by JarnoRFB** [comment]
Given the widespread use of `attrs` and the fact that `dataclasses` are not a [full replacement](http://www.attrs.org/en/stable/why.html#data-classes) of `attrs` I would still argue that type checking support from `pyright` would be justified. Even `mypy` ships with [dedicated support](http://www.attrs.org/en/stable/types.html#type-annotations) for `attrs`, so I would not call it an *arbitrary* third-party library.

**Comment by ayroblu** [comment]
Hey just wondering if there's any reason for pyright over mypy?

**Comment by layday** [comment]
You can inherit from a protocol to silence errors in Pyright:

```py
from typing import Any, Protocol

import attr


class AttrProto(Protocol):
    def __init__(self, **kwargs: Any) -> None: ...


@attr.s()
class Foo(AttrProto):
    bar: str = attr.ib()


Foo(bar='baz')   # ok
```

Modifying the attrs stub file to use a generic protocol would also work.

**Comment by Hugovdberg** [comment]
As mentioned before, `attrs` is quite widely used ([~24 million downloads last month](https://pypistats.org/packages/attrs)), are there any plans to support this "arbitrary third-party" package or is work necessary from the @python-attrs developers?

**Comment by erictraut** [comment]
We have no plans for Pyright to include knowledge of third-party libraries (i.e. those that are not part of the Python specification). Third-party libraries must be described by type stub files.

We might eventually provide an extensibility mechanism so Pyright's type logic can be extended by plug-ins, but such an extension mechanism would require significant work, add significant internal complexity, and limit future changes to Pyright. So it's not something we are in a rush to provide. Pyright needs to mature a bit more first.

**Comment by Hugovdberg** [comment]
I can understand that, but that would mean that Pyright is currently not very usable when you depend on `attrs` or for example `SQLAlchemy`. But I can understand you want to let Pyright mature first before making things a lot more complex 👍

**Comment by glyph** [comment]
> dataclass is part of the official python spec

For what it's worth, [Python does not have a language specification](https://stackoverflow.com/questions/1094961/is-there-a-python-language-specification), let alone an "official" one, and if it did, `dataclasses` would not be part of it, since the closest thing is the [language reference](https://docs.python.org/3/reference/) which only describes `importlib`.  I think you mean "part of the CPython standard library distribution", since there are many Python runtimes and not all of them include `dataclasses`.

This might seem like quibbling about semantics, but the nature of Python's language evolution is that packages like `attrs` frequently evolve outside the stdlib and are adopted by consensus of the community rather than being dictated by formal specification updates.

**Comment by erictraut** [comment]
When I say "official Python spec", I'm referring to PEPs like [PEP 557](https://www.python.org/dev/peps/pep-0557/). These PEPs go through a formal and rigorous process of review and ratification. Once adopted, there is a strong guarantee of compatibility (and backward compatibility going forward).

**Comment by orsinium** [comment]
> We might eventually provide an extensibility mechanism so Pyright's type logic can be extended by plug-ins

A better alternative would be to support MyPy plugins, even if it would require dirty hacks and mocking mypy public interfaces. There are already plenty of them (see [awesome-python-typing](https://github.com/typeddjango/awesome-python-typing)), it would be great to be able to reuse them rather than re-implement. For instance, [attrs support is implemented as a plugin](https://github.com/python/mypy/blob/master/mypy/plugins/attrs.py).

**Comment by erictraut** [comment]
We've looked into the possibility of supporting mypy plugins, but they're too tied to the internal object model of mypy.

**Comment by asford** [comment]
The upcoming dataclass transforms are a potentially interesting follow-up to anyone else who encounters this issue:

https://github.com/microsoft/pyright/blob/master/specs/dataclass_transforms.md

https://github.com/microsoft/pyright/pull/1773

**Comment by erictraut** [comment]
@asford, thanks for that comment. I posted a note in the companion issue in the pylance repo (https://github.com/microsoft/pylance-release/issues/226) but forgot to post here as well.

We have been exploring ways to provide better support for `attrs` and other libraries that support semantics similar to `dataclass`. Pylance version 2021.4.2 (released earlier today) contains a proposed solution that is documented in [this spec](https://github.com/microsoft/pyright/blob/master/specs/dataclass_transforms.md#applying-to-attrs). If you want to try it, I've included instructions on how to modify your local copy of `attrs` (or `pydantic`) source code. We welcome early feedback.

I've started a [discussion in the pyright repo](https://github.com/microsoft/pyright/discussions/1782). Please post questions, comments, or feedback there.
